src/WorkflowStep.ts

```diff
@@ -0,0 +1,379 @@
+    throw new WorkflowStepInitializationError(errorMsg);
+  }
+
+  // Ensure a callback or an array of callbacks is present
+  const requiredFns: (keyof WorkflowStepOptions)[] = ['save', 'edit', 'execute'];
+  requiredFns.forEach((fn) => {
+    if (typeof config[fn] !== 'function' && !Array.isArray(config[fn])) {
+      const errorMsg = `WorkflowStep ${fn} property must be a function or an array of functions`;
+      throw new WorkflowStepInitializationError(errorMsg);
+    }
+  });
+}
+
+/**
+ * `processStepMiddleware()` invokes each callback for lifecycle event
+ * @param args workflow_step_edit action
+ */
+export async function processStepMiddleware(
+  args: AllWorkflowStepMiddlewareArgs,
+  middleware: WorkflowStepMiddleware,
+): Promise<void> {
+  const { context, client, logger } = args;
+  // TODO :: revisit type used below (look into contravariance)
+  const callbacks = [...middleware] as Middleware<AnyMiddlewareArgs>[];
+  const lastCallback = callbacks.pop();
+
+  if (lastCallback !== undefined) {
+    await processMiddleware(callbacks, args, context, client, logger, async () =>
+      lastCallback({ ...args, context, client, logger }),
+    );
+  }
+}
+
+export function isStepEvent(args: AnyMiddlewareArgs): args is AllWorkflowStepMiddlewareArgs {
+  const validTypes = new Set(['workflow_step_edit', 'workflow_step', 'workflow_step_execute']);
+  return validTypes.has(args.payload.type);
+}
+
+function selectToken(context: Context): string | undefined {
+  return context.botToken !== undefined ? context.botToken : context.userToken;
+}
+
+/**
+ * Factory for `configure()` utility
+ * @param args workflow_step_edit action
+ */
+function createStepConfigure(
+  args: AllWorkflowStepMiddlewareArgs<SlackActionMiddlewareArgs<WorkflowStepEdit>>,
+): StepConfigureFn {
+  const {
+    context,
+    client,
+    body: { callback_id, trigger_id },
+  } = args;
+  const token = selectToken(context);
+
+  return (config: Parameters<StepConfigureFn>[0]) => {
+    const { blocks, private_metadata, submit_disabled, external_id } = config;
+    return client.views.open({
+      token,
+      trigger_id,
+      view: {
+        callback_id,
+        blocks,
+        type: 'workflow_step',
+        // only include private_metadata, submit_disabled + external_id if passed by user
+        ...(private_metadata !== undefined && { private_metadata }),
+        ...(submit_disabled !== undefined && { submit_disabled }),
+        ...(external_id !== undefined && { external_id }),
+      },
+    });
+  };
+}
+
+/**
+ * Factory for `update()` utility
+ * @param args view_submission event
+ */
+function createStepUpdate(
+  args: AllWorkflowStepMiddlewareArgs<SlackViewMiddlewareArgs<ViewWorkflowStepSubmitAction>>,
+): StepUpdateFn {
+  const {
+    context,
+    client,
+    body: {
+      workflow_step: { workflow_step_edit_id },
+    },
+  } = args;
+  const token = selectToken(context);
+
+  return (config: Parameters<StepUpdateFn>[0]) => {
```

Comment:
would it be worth adding a default value to handle cases where the developer passes nothing in. Technically it will work without any of those values passed in (useful for setting up a hello-world step that doesn't actually do much yet and is static). Having a default of an empty object here, and for the `complete` callback would solve that use case.

Reply:
Whoops, I dropped this on the wrong place, I meant for the `update` and `complete` calls